arch/risc-v/src/common/supervisor/riscv_context.S

```diff
@@ -0,0 +1,234 @@
+
+  REGLOAD    s0, REG_EPC(sp)           /* restore epc */
+  csrw       CSR_EPC, s0
+
+  /* Restore status register, but don't enable interrupts yet */
+
+  REGLOAD    s0, REG_INT_CTX(sp)       /* restore status */
+  li         s1, STATUS_IE             /* move IE -> PIE */
+  and        s1, s0, s1                /* if (STATUS & IE) */
+  beqz       s1, 1f
+  li         s1, ~STATUS_IE            /* clear IE */
+  and        s0, s0, s1
+  li         s1, STATUS_PIE            /* set PIE */
+  or         s0, s0, s1
+
+1:
+  csrw       CSR_STATUS, s0
+
+  load_ctx   sp
+
+  REGLOAD    sp, REG_SP(sp)            /* restore original sp */
+
+  /* return from exception, which updates the status register */
+
+  sret
+
+/****************************************************************************
+ * Name: riscv_switchcontext
+ *
+ * Description:
+ *   Restore user context
+ *
+ * C Function Prototype:
+ *   void riscv_switchcontext(uintptr_t *saveregs, uintptr_t *restoreregs);
+ *
+ * Input Parameters:
+ *   saveregs    - Context to save
+ *   restoreregs - Context to restore
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Global interrupts disabled by the caller.
+ *
+ ****************************************************************************/
+
+.type riscv_switchcontext, function
+
+riscv_switchcontext:
+
+  /* Save old context to arg[0] */
+
+  save_ctx   a0                        /* save context */
+
+  REGSTORE   x1, REG_EPC(a0)           /* save ra to epc */
+  REGSTORE   sp, REG_SP(a0)            /* original SP */
+
+  /* Set previous privilege, we are in privileged mode now */
+
+  csrr       s0, CSR_STATUS            /* read status register */
+  li         s1, STATUS_PPP            /* set previous privilege */
+  or         s0, s0, s1
+  li         s1, ~STATUS_PIE           /* clear previous interrupt enable */
+  and        s0, s0, s1
+  REGSTORE   s0, REG_INT_CTX(a0)       /* store status to context */
+
+#ifdef CONFIG_ARCH_FPU
+  jal        x1, riscv_savefpu         /* FP registers */
+#endif
+
+  /* Load new context from arg[1] */
+
+  mv         a0, a1                    /* load from a1 */
+  j          riscv_fullcontextrestore  /* restore context */
+
+/****************************************************************************
+ * Name: riscv_syscall_return
+ *
+ * Description:
+ *   Return from system call to user task with user mode privileges
+ *
+ * C Function Prototype:
+ *   void riscv_syscall_return(void);
+ *
+ * Input Parameters:
+ *   Assumes the return value of the system call is in a0
+ *
+ * Returned Value:
+ *   Return value of system call is returned into contex
+ *
+ * Assumptions:
+ *   User task is running system call in privileged mode, ready to resume
+ *   the user task in unprivileged mode
+ *
+ ****************************************************************************/
+
+.type riscv_syscall_return, function
+
+riscv_syscall_return:
```

Comment:
One exception is riscv_syscall_return, this function disables interrupts before jumping, which is a must if the trampoline function is used. The other functions are called with interrupts disabled always, so no need to mask interrupts explicitly.

When using ecall (in riscv_swint.c) IRQ:s are masked automatically, so for that no special functionality is needed. But if a jump is used, IRQ:s must be masked. I guess that could be done in the dispatch function also, but we would have to move every line of code from this asm function there.

Reply:
riscv_syscall_dispatch should simulate ECALL fully which mean:

1. Ensure interrupt is disabled before call  riscv_handle_syscall
2. Restore the interrupt state before return

Since this is the most trick part and may require the fine tune with more test, that's why I suggest all syscall include context switch is better to go through the same path, so we can verify and adapter the code in one place.

Reply:
Very well. I'll move the IRQ masking there.

Reply:
Done